# Business and Data Understanding

This notebook documents the **Business Understanding** and **Data Understanding** phases of the CRISP-DM methodology for the *Mental Health in Technology* project.

The analysis uses the OSMI Mental Health in Tech 2016 survey dataset and establishes the project's analytical context, objectives, success criteria, and understanding of the raw data before any data preparation or modelling is performed.

## 1. Business Understanding

Before examining the dataset, this section establishes the project context, analytical objectives, data-mining goals, and criteria for evaluating the eventual clustering results.

The project follows the **CRISP-DM (Cross-Industry Standard Process for Data Mining)** methodology. At this stage, the focus is limited to Business Understanding and Data Understanding. No data cleaning, transformation, feature engineering, dimensionality reduction, or clustering is performed in this notebook.

### 1.1 Project Background

The project analyses the **OSMI Mental Health in Tech 2016 survey** to investigate patterns associated with mental health, workplace conditions, treatment-seeking behaviour, and disclosure among survey respondents.

The study is designed as an **unsupervised machine learning problem**. There is no predefined target variable that the models are required to predict. Instead, the analysis seeks to identify groups of respondents who exhibit similar characteristics across the variables selected for the subsequent analytical stages.

The resulting groups will be interpreted as analytical segments of the survey population. They will not be treated as clinical diagnoses, psychological classifications, or measures of individual mental-health risk.

### 1.2 Business Objective

The primary objective is to investigate whether the survey data contains meaningful respondent segments associated with different patterns of:

- Mental health conditions and experiences
- Treatment-seeking behaviour
- Workplace mental-health support
- Perceived workplace safety and stigma
- Mental health disclosure
- Demographic and employment characteristics

The resulting analysis may provide evidence that helps describe differences in workplace mental-health experiences represented within the survey.

The study is exploratory. Consequently, identified associations should not be interpreted as causal relationships, clinical diagnoses, or individual-level risk assessments.

### 1.3 Data Mining Goals

The data-mining process will aim to:

1. Understand the structure, content, and quality characteristics of the raw survey data.
2. Examine the distributions and missing-data patterns present in the available variables.
3. Identify variables that are potentially relevant to the project's unsupervised-learning objective.
4. Prepare an appropriate analytical representation of the survey data.
5. Apply dimensionality-reduction techniques where appropriate.
6. Apply multiple clustering approaches to investigate potential respondent segments.
7. Evaluate clustering solutions using appropriate quantitative and qualitative evidence.
8. Interpret the resulting clusters using the original survey variables and project objectives.

The clustering analysis will consider the following approaches:

- **K-Means**
- **Gaussian Mixture Models (GMM)**
- **Hierarchical Clustering**
- **DBSCAN**

Dimensionality-reduction and feature-selection methods will be considered where appropriate to support the subsequent unsupervised-learning analysis.

### 1.4 Success Criteria

The project will be considered analytically successful if the resulting clustering analysis provides solutions that:

- Demonstrate measurable structure or separation within the analysed feature space.
- Can be evaluated using appropriate quantitative clustering measures.
- Exhibit sufficient interpretability to support meaningful examination of the resulting groups.
- Can be related back to identifiable characteristics in the original survey variables.
- Provide descriptive insight into patterns of mental-health experiences, workplace conditions, and related behaviours.
- Are supported by quantitative and qualitative evidence rather than visual inspection alone.

Any resulting clusters will be treated as **analytical segments of the survey data**, rather than clinical, psychological, or diagnostic classifications.

## 2. Data Understanding

The Data Understanding phase begins with the original survey data before any cleaning, transformation, feature engineering, dimensionality reduction, or clustering is performed.

The raw dataset is loaded using the reusable data-ingestion utility implemented in:

`src/mental_health_ml/data.py`

Separating data ingestion from notebook analysis keeps the notebook focused on the analytical workflow while providing a reusable and consistent mechanism for loading the project's raw dataset.

In [46]:
# import warning

# Import the standard-library module for filesystem path handling.
from pathlib import Path

# Import pandas for tabular data inspection and analysis.
import pandas as pd

# Import the project's reusable dataset-loading utility.
from mental_health_ml.data import load_dataset

# warnings.filterwarnings("ignore")
# I decided to comment out the global warning suppression becasuse
# warnings can contain useful information about data-loading or
# analytical behaviour.

### 2.1 Raw Dataset

The dataset is loaded directly from the project's raw-data directory.

The raw file is retained separately from subsequent analytical outputs so that the original survey data remains available as the source dataset for the project.

In [47]:
# Define the location of the immutable raw survey archive.
RAW_DATA_PATH = Path("../data/raw/mental-heath-in-tech-2016_20161114.csv.zip")

# Display the configured path so that the notebook records which raw
# dataset location is being used for the analysis.
RAW_DATA_PATH

WindowsPath('../data/raw/mental-heath-in-tech-2016_20161114.csv.zip')

### 2.2 Dataset Ingestion

The reusable `load_dataset()` function is used to load the survey data. The loader handles the ZIP archive and reads the CSV contained within it without modifying the raw observations.

In [48]:
# Load the original survey dataset using the project's reusable
# data-ingestion utility.
df = load_dataset(RAW_DATA_PATH)

# Confirm the resulting object type before beginning inspection.
type(df)

pandas.DataFrame

### 2.3 Initial Record Inspection

The first and last observations are inspected to establish the structure and content of the records at the beginning and end of the raw dataset.

This inspection is descriptive only. No values are modified at this stage.

In [49]:
# Inspect the first five observations to verify that the dataset
# has been loaded correctly and that the columns contain expected data.
df.head()

,Are you self-employed?,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health concerns and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,"If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:",...,"If you have a mental health issue, do you feel that it interferes with your work when being treated effectively?","If you have a mental health issue, do you feel that it interferes with your work when NOT being treated effectively?",What is your age?,What is your gender?,What country do you live in?,What US state or territory do you live in?,What country do you work in?,What US state or territory do you work in?,Which of the following best describes your work position?,Do you work remotely?
0,0,26-100,1.0,NaN,Not eligible for coverage / N/A,NaN,No,No,I don't know,Very easy,...,Not applicable to me,Not applicable to me,39,Male,United Kingdom,NaN,United Kingdom,NaN,Back-end Developer,Sometimes
1,0,6-25,1.0,NaN,No,Yes,Yes,Yes,Yes,Somewhat easy,...,Rarely,Sometimes,29,male,United States of America,Illinois,United States of America,Illinois,Back-end Developer|Front-end Developer,Never
2,0,6-25,1.0,NaN,No,NaN,No,No,I don't know,Neither easy nor difficult,...,Not applicable to me,Not applicable to me,38,Male,United Kingdom,NaN,United Kingdom,NaN,Back-end Developer,Always
3,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Sometimes,Sometimes,43,male,United Kingdom,NaN,United Kingdom,NaN,Supervisor/Team Lead,Sometimes
4,0,6-25,0.0,1.0,Yes,Yes,No,No,No,Neither easy nor difficult,...,Sometimes,Sometimes,43,Female,United States of America,Illinois,United States of America,Illinois,Executive Leadership|Supervisor/Team Lead|Dev ...,Sometimes


In [50]:
# Inspect the final five observations to verify the end of the
# raw dataset and confirm that the complete table has been loaded.
df.tail()

,Are you self-employed?,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health concerns and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,"If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:",...,"If you have a mental health issue, do you feel that it interferes with your work when being treated effectively?","If you have a mental health issue, do you feel that it interferes with your work when NOT being treated effectively?",What is your age?,What is your gender?,What country do you live in?,What US state or territory do you live in?,What country do you work in?,What US state or territory do you work in?,Which of the following best describes your work position?,Do you work remotely?
1428,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Not applicable to me,Not applicable to me,34,Female,United States of America,New York,United States of America,New York,Other,Sometimes
1429,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Sometimes,Often,56,MALE,United States of America,California,Afghanistan,NaN,Support,Sometimes
1430,0,100-500,1.0,NaN,Yes,Yes,Yes,Yes,I don't know,Somewhat difficult,...,Rarely,Sometimes,52,Male,United States of America,Georgia,United States of America,Georgia,Back-end Developer,Sometimes
1431,0,100-500,0.0,1.0,I don't know,I am not sure,No,Yes,I don't know,Somewhat difficult,...,Sometimes,Often,30,Female,United States of America,Nebraska,United States of America,Nebraska,DevOps/SysAdmin,Sometimes
1432,0,100-500,1.0,NaN,Yes,No,No,No,I don't know,Very difficult,...,Not applicable to me,Often,25,non-binary,Canada,NaN,Canada,NaN,Other,Sometimes


### 2.4 Dataset Dimensions

The dataset dimensions are examined to establish the number of observations and variables available for the project.

In [52]:
# Report the number of observations and variables in the raw dataset.
print(f"The dataset contains {df.shape[0]:,} observations and {df.shape[1]} variables.")

The dataset contains 1,433 observations and 63 variables.


### 2.5 Column Names

The complete set of column names is inspected to understand the information captured by the survey and to identify the broad subject areas represented in the raw data.

The column names are retained exactly as supplied by the source dataset at this stage.

In [53]:
# Display the complete list of raw survey variables.
df.columns.tolist()

['Are you self-employed?',
 'How many employees does your company or organization have?',
 'Is your employer primarily a tech company/organization?',
 'Is your primary role within your company related to tech/IT?',
 'Does your employer provide mental health benefits as part of healthcare coverage?',
 'Do you know the options for mental health care available under your employer-provided coverage?',
 'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?',
 'Does your employer offer resources to learn more about mental health concerns and options for seeking help?',
 'Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?',
 'If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:',
 'Do you think that discussing a mental health disorder with your employer would have neg

### 2.6 Data Types

The pandas data types are inspected to determine the current representation of each raw variable.

This inspection is important because the survey contains a mixture of numerical and text-based responses. The data types observed here describe the raw representation only; no conversion is performed in this notebook.

In [54]:
# Inspect the pandas data type assigned to each raw survey variable.
df.dtypes

Are you self-employed?                                                                 int64
How many employees does your company or organization have?                               str
Is your employer primarily a tech company/organization?                              float64
Is your primary role within your company related to tech/IT?                         float64
Does your employer provide mental health benefits as part of healthcare coverage?        str
                                                                                      ...   
What US state or territory do you live in?                                               str
What country do you work in?                                                             str
What US state or territory do you work in?                                               str
Which of the following best describes your work position?                                str
Do you work remotely?                                                 

### 2.7 Structural Dataset Information

The `DataFrame.info()` summary provides a consolidated view of:

- The number of observations
- The number of variables
- Non-null observations per variable
- Current data types
- Approximate memory usage

This provides the first systematic view of the completeness and structural composition of the raw dataset.

In [55]:
# Display the consolidated structural summary of the raw dataset.
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1433 entries, 0 to 1432
Data columns (total 63 columns):
 #   Column                                                                                                                                                                            Non-Null Count  Dtype  
---  ------                                                                                                                                                                            --------------  -----  
 0   Are you self-employed?                                                                                                                                                            1433 non-null   int64  
 1   How many employees does your company or organization have?                                                                                                                        1146 non-null   str    
 2   Is your employer primarily a tech company/organization?                

### 2.8 Missing-Value Inspection

The raw dataset is inspected for missing observations before any decisions are made about their treatment.

Two measures are calculated:

- **Missing count:** the number of null observations for each variable.
- **Missing percentage:** the proportion of observations that are null for each variable.

No missing values are imputed, removed, or otherwise transformed in this notebook.

The purpose is to establish the extent and distribution of missing information so that appropriate preparation decisions can be made in the subsequent Data Preparation phase.

In [56]:
# Count missing observations for every variable and order the results
# from the highest number of missing observations to the lowest.
missing_values = df.isna().sum().sort_values(ascending=False)

# Calculate the percentage of observations missing from each variable.
missing_percentage = df.isna().mean().mul(100).sort_values(ascending=False)

# Combine the missing-value count and percentage into a single
# inspection table for easier interpretation.
missing_summary = pd.DataFrame(
    {
        "missing_count": missing_values,
        "missing_percentage": missing_percentage,
    }
)

# Display the 20 variables with the highest number of missing observations.
missing_summary.head(20)

,missing_count,missing_percentage
"If you have revealed a mental health issue to a client or business contact, do you believe this has impacted you negatively?",1289,89.951151
"If yes, what percentage of your work time (time performing primary or secondary job functions) is affected by a mental health issue?",1229,85.764131
Is your primary role within your company related to tech/IT?,1170,81.646895
Do you have medical coverage (private insurance or state-provided) which includes treatment of mental health issues?,1146,79.972087
"If you have been diagnosed or treated for a mental health disorder, do you ever reveal this to clients or business contacts?",1146,79.972087
"If you have been diagnosed or treated for a mental health disorder, do you ever reveal this to coworkers or employees?",1146,79.972087
"If you have revealed a mental health issue to a coworker or employee, do you believe this has impacted you negatively?",1146,79.972087
Do you know local or online resources to seek help for a mental health disorder?,1146,79.972087
Do you believe your productivity is ever affected by a mental health issue?,1146,79.972087
"If maybe, what condition(s) do you believe you have?",1111,77.529658


In [57]:
# Display the 20 variables with the lowest number of missing observations
missing_summary.tail(20)

,missing_count,missing_percentage
What is your gender?,3,0.209351
Are you self-employed?,0,0.000000
Do you have previous employers?,0,0.000000
Would you be willing to bring up a physical health issue with a potential employer in an interview?,0,0.000000
Have you had a mental health disorder in the past?,0,0.000000
Do you have a family history of mental illness?,0,0.000000
Do you think that team members/co-workers would view you more negatively if they knew you suffered from a mental health issue?,0,0.000000
How willing would you be to share with friends and family that you have a mental illness?,0,0.000000
Do you feel that being identified as a person with a mental health issue would hurt your career?,0,0.000000
Would you bring up a mental health issue with a potential employer in an interview?,0,0.000000


### Missing-Value Observations

The missingness is substantial but unevenly distributed across the survey variables.

Several variables contain very high levels of missingness. For example, the question concerning whether the respondent's primary role is related to technology/IT has 263 non-null observations, while several mental-health-related questions contain 287 non-null observations. In contrast, variables such as age, country of residence, country of work, work position, and remote-working status contain observations for all respondents.

The pattern indicates that missing values are not distributed uniformly across the questionnaire. Some questions appear to have been presented or answered conditionally, meaning that a missing value does not necessarily represent the same situation as an unanswered question.

### 2.9 Duplicate-Observation Inspection

The raw dataset is checked for completely duplicated rows.

This inspection determines whether multiple observations contain identical values across all 63 variables. No rows are removed as part of this check because this notebook is limited to understanding the raw dataset.

In [58]:
# Count rows that are exact duplicates of another complete observation.
duplicate_count = df.duplicated().sum()

if duplicate_count == 0:
    print("There are no duplicated observations.")
else:
    print(f"The dataset contains {duplicate_count:,} duplicated rows.")

There are no duplicated observations.


The duplicate check identifies no completely duplicated observations in the raw dataset.

### 2.10 Unique-Value Inspection

The number of unique values in each variable is inspected to understand the cardinality of the survey responses.

Missing values are included in the count because they represent an observed characteristic of the raw dataset at this stage.

This inspection is particularly useful for distinguishing low-cardinality survey responses from higher-cardinality demographic and free-text variables.

In [59]:
# Count distinct values for every variable, including missing values,
# and order the variables from highest to lowest cardinality
unique_value_counts = df.nunique(dropna=False).sort_values(ascending=False)

# Display the 20 variables with the highest number of disntict values.
unique_value_counts.head(20)

Why or why not?                                                                                                                   1086
Why or why not?.1                                                                                                                 1081
Which of the following best describes your work position?                                                                          264
If yes, what condition(s) have you been diagnosed with?                                                                            129
If so, what condition(s) were you diagnosed with?                                                                                  117
If maybe, what condition(s) do you believe you have?                                                                               100
What is your gender?                                                                                                                71
What country do you live in?                           

In [60]:
# Display the 20 variables with the lowest number of distinct values.
unique_value_counts.tail(20)

Do you feel that your employer takes mental health as seriously as physical health?                                                                                                 4
If you have revealed a mental health issue to a client or business contact, do you believe this has impacted you negatively?                                                        4
Does your employer offer resources to learn more about mental health concerns and options for seeking help?                                                                         4
Do you know the options for mental health care available under your employer-provided coverage?                                                                                     4
Have your observations of how another individual who discussed a mental health disorder made you less likely to reveal a mental health issue yourself in your current workplace?    4
Would you have been willing to discuss a mental health issue with your previous co-workers

The cardinality inspection shows substantial variation across the survey variables.

The two free-text `"Why or why not?"` variables have more than 1,000 distinct values, while work-position and demographic variables also have relatively high cardinality. In contrast, many structured survey questions contain only a small number of response categories.

This confirms that the raw dataset contains a mixture of structured categorical responses, demographic variables, and free-text responses.

### 2.11 Representative Categorical-Variable Inspection

The raw dataset contains a large number of text-based variables. A representative subset is therefore inspected to understand the response categories and their frequencies.

This is an exploratory inspection rather than a transformation step. The original values are preserved exactly as loaded.

In [61]:
# Identify variables currently represented by pandas as text, categorical,
# or Boolean data types.
categorical_columns = df.select_dtypes(include=["object", "category", "bool"]).columns

print(f"Categorical/text variables identified in the raw dataset: {len(categorical_columns):,}")

Categorical/text variables identified in the raw dataset: 56


In [62]:
# Display the frequency distribution for the first ten categorical/text
# variables as a representative sample of the raw response structure.
for column in categorical_columns[:10]:
    print(f"\n{column}")
    print(df[column].value_counts(dropna=False).head(10))


How many employees does your company or organization have?
How many employees does your company or organization have?
26-100            292
NaN               287
More than 1000    256
100-500           248
6-25              210
500-1000           80
1-5                60
Name: count, dtype: int64

Does your employer provide mental health benefits as part of healthcare coverage?
Does your employer provide mental health benefits as part of healthcare coverage?
Yes                                531
I don't know                       319
NaN                                287
No                                 213
Not eligible for coverage / N/A     83
Name: count, dtype: int64

Do you know the options for mental health care available under your employer-provided coverage?
Do you know the options for mental health care available under your employer-provided coverage?
NaN              420
No               354
I am not sure    352
Yes              307
Name: count, dtype: int64

Has your em

The representative frequency tables demonstrate that many survey variables use qualitative response categories rather than simple binary values.

For example, employer mental-health-benefit responses include categories such as `"Yes"`, `"No"`, `"I don't know"`, and `"Not eligible for coverage / N/A"`. Other workplace questions contain categories such as `"Yes"`, `"No"`, `"Maybe"`, `"I don't know"`, or graded responses relating to difficulty or comfort.

The raw response categories therefore require careful consideration during the subsequent preparation stage.

## 3. Data Understanding Summary

The Data Understanding phase establishes the following characteristics of the raw OSMI Mental Health in Tech 2016 survey dataset:

- The dataset contains **1,433 observations and 63 variables**.
- The raw data is predominantly text-based, with **56 text variables**, alongside numerical variables represented by integer and floating-point data types.
- Missing observations are unevenly distributed across the questionnaire, with some variables containing substantial missingness and others being complete.
- Several variables have high cardinality, particularly free-text and demographic variables, while many structured survey questions have relatively few response categories.
- No completely duplicated observations were identified.
- The survey contains a mixture of workplace, demographic, mental-health, treatment, disclosure, and employment-related information.
- The missing-data patterns and mixed variable representations indicate that the raw dataset requires deliberate preparation before it can be used for unsupervised machine learning.

These findings define the starting point for the next CRISP-DM stage.

**No data cleaning, missing-value treatment, categorical encoding, feature engineering, dimensionality reduction, or clustering has been performed in this notebook.**

The next notebook will address **Data Preparation and Feature Engineering**, where the raw variables will be assessed and transformed into an appropriate analytical representation for the subsequent dimensionality-reduction and clustering stages.